### PR TITLE
Roman PSF for Ensemble Test

### PR DESCRIPTION
**Pull request by sweverett**
## Summary

Wires the realistic Roman WFI PSF (`galsim.roman.getPSF`) into the ensemble production path as a config-boundary change; the render chain was already GSObject-agnostic. Gaussian configs are fully backward compatible.

This PR was done on a separate worktree from `se/speedups` to implement this infrastructure in parallel with the ensemble/GPU work. It merges into `se/speedups`, not `main`; no outside review needed.

- New PSFSpec config type `roman_wfi` (per-channel; `sca` default 10, `pupil_bin` default 4, optional `folding_threshold`); mixed gaussian/roman configs allowed; unknown types still raise.
- Configs use official Roman filter names (F158, F184); galsim's legacy WFIRST keys (H158, Z087, ...) are mapped at the galsim boundary and still accepted.
- Broadband kernels rendered monochromatic at the band effective wavelength (F158 1579 nm, F184 1842 nm; galsim 2.8.4, getPSF takes nm).
- Grism kernel rendered at the observed Halpha wavelength 656.28(1+z) nm per fit, with the kernel array shape pinned at the spec's z-draw upper bound so shapes are constant across the ensemble (persistent-compile-cache friendly). Pinning = zero-padding each fit's own correct kernel; never a worst-case PSF.
- `kernel_size` override in psf.py/observation.py: odd, >= GalSim good size, raises rather than truncate; preserved through `with_render_config` (obs field).
- Effective-maxk scans cap the PSF factor at `psf.maxk`: the interpolated-pupil PSF's kValue sidelobes exceed the 1e-3 threshold beyond the physical aperture cutoff, which inflated the priors-implied oversample to 7 and made `InferenceTask.from_obs` raise on the production worker path (found by audit; regression-tested).
- Process-level getPSF cache keyed (sca, pupil_bin, wavelength@0.1nm, bandpass, folding_threshold), FIFO-capped at 16.

## Measured cost + the folding_threshold knob

macbook CPU, canonical_P_roman vs canonical_P (gaussian), oversample 3, z=1.2, kernel pinned at z=1.9, logpost+grad eval median:

| config | grism padded FFT | eval | vs gaussian |
|---|---|---|---|
| gaussian canonical_P | 125^2 | 37.5 ms | 1.0x |
| roman, default kernels | 1089^2 | 141.6 ms | 3.8x |
| roman, folding_threshold 0.01 | 600^2 | 51.0 ms | 1.4x |
| roman, folding_threshold 0.02 | 350^2 | 35.0 ms | 0.9x |

Noiseless-render distortion vs default kernels at line SNR 100 (whole-stamp chi): 0.045 at ft=0.01, 0.51 at ft=0.02; broadband path unaffected (k-space analytic kernel FT, pinned by exact-equality test). Compile time unchanged. The knob is config-driven and NOT enabled in `canonical_P_roman.yaml`; production choice pending. Fidelity gate test bounds = 10x measured (provenance in test docstring).

## Deliberately out of scope

Flagship test migration to Roman PSF (frozen refs move; follow-up after this PR lands, with user consultation), chromatic/per-slice PSFs (issue #51), SCA-position dependence, ImCom coadd PSFs. Note: grism kernel-size pinning currently requires a uniform z draw; the planned LF-based n(z) sampler will hit a loud NotImplementedError there (extension scheduled with the population-sampler work).

## Review focus

**API:**
- `kl_pipe/ensemble/spec.py` -- PSFSpec dataclass + YAML schema (incl. folding_threshold)
- `kl_pipe/psf.py`, `kl_pipe/observation.py` -- kernel_size pinning semantics (Critical: refuses to truncate; requires explicit render_config when pinned)
- `kl_pipe/render.py` -- effective-maxk PSF band-limit cap (Critical: grid sizing)

**Internals:**
- `kl_pipe/ensemble/mocks.py` -- PSF builder dispatch, band-name map, getPSF cache, z-dependent grism kernel, pin-size helper

**Tests:**
- `tests/test_roman_psf.py` (37: parsing, kernel physics, shape pinning, sidelobe-cap regression, from_obs production path, folding fidelity gate)
- `tests/test_ensemble.py`, `tests/test_render_config.py` (accessor updates; _DeadPSF fake gains maxk)
- test_flagship and all frozen references untouched

Verified at 4a7e68a: test_roman_psf 34 (31 fast + 3 slow) passed; test_ensemble + test_render_config passed (107 fast total); earlier full adjacent sweep at 91f104f: 532 passed / 0 failed. make check-format clean (pre-commit black).

---

## Addendum (through e4b522c): rigor tests, z-tiered folding_threshold, rfft2 + minimal-padding swap

Five commits added after the original review round. All additive except one test-content pin refresh (flagged below).

**Rigor tests (123bad2).** Kernel-convention checks with the real (asymmetric) roman kernel: k-space fused path vs galsim native 7.0e-4 max|diff|/peak; a deliberately transposed kernel reads 8.0e-3 and the 2e-3 bound sits between, so a parity/orientation bug cannot pass. Real-space delta-patch identity 7.1e-8; kernel DC == 1; coarse pixel applied exactly once post-dispersion. maxk cap re-verified: psf.maxk = 0.992x the analytic aperture cutoff at z=0.55/1.2/1.9.

**z-tiered fit folding_threshold (25b6b00, d6940b9).** The z=1.2 dual-fidelity acceptance of ft=0.01 does not extend to high z: mock-vs-fit MAP bias grows with kernel size (cosi -0.093 sigma, g1 -0.037, continuum -0.066 at z=1.9). Fix: `PSFSpec.folding_threshold_tiers` (mutually exclusive with the scalar; loud validation; mock >= tightest tier enforced), and canonical_P_roman.yaml now runs ft=0.01 for z<=1.2, ft=5e-3 above. At the tight tier the fit kernel is bit-identical to the mock kernel (max|diff| = 0), so the mismatch is zero by construction. Cost at z=1.9 (interleaved A/B; ratios robust, absolutes contention-contaminated):

| psf.py | ft=0.01 | ft=5e-3 | tight/loose |
|---|---|---|---|
| pre-swap | 48.4 ms (600^2 grism FFT) | 134.0 ms (1089^2) | 2.77x |
| with the swap below | 37.0 ms (350^2) | 57.7 ms (594^2) | 1.56x |

Ensemble-mean multiplier at f(z>1.2) ~= 0.49 (population-proto selected n(z)): 1.87x pre-swap, 1.27x with the swap.

**rfft2 + minimal-exact padding (5c6f90c).** Single production path in psf.py, no config knob. FFT convolution is circular mod the padded size P; with kernel half-width H = K//2, wraparound can reach a retained pixel only if P < N + H, so P >= N + H is exact on the retained N x N crop -- the grid shrinks from N+K-1 to N+K//2 (roman wings make K >> N, hence the win). Kernel support is finite by construction (folding_threshold-sized drawImage stamp). Old path kept verbatim in tests as `_reference_full_pad_fft2_convolve`; direct old-vs-new gate over {symmetric, asymmetric} kernels x 4 grid parities: worst-case rel delta 5.6e-16 value, 6.1e-16 grad. Bound 1e-14 = float64-identity floor, not a physics tolerance. Diagnostic panel regenerates to tests/out/psf_rfft2_swap/.

**Existing-test change (e4b522c, the only one).** test_registry_roman_config_loads pinned the yaml's old scalar ft=0.01; refreshed to pin the full tier schedule + scalar None + mock unset -- strictly more constrained than before. No tolerance changed anywhere.

**Validation at e4b522c** (local; via python -m pytest): targeted psf/roman/ensemble/galsim/cube/grism suites 214 passed; fast tier 1116 passed / 2 skipped; flagship dev joint Nsigma 0.086, production 0.845 (gates < 3). Both flagship variants build PSFSpec inline, so they exercise the swap but not the yaml tiers -- tier coverage is unit-level (TestFoldingThresholdTiers, 12 tests) plus the registry pin; an e2e yaml-tier test is a follow-up.

**Review focus (additions):**

- (Critical) `kl_pipe/psf.py` -- the padding argument above is the correctness crux
- `kl_pipe/ensemble/spec.py` -- FoldingThresholdTier schedule + validation
- `configs/observing/canonical_P_roman.yaml` -- tier flip
- `kl_pipe/ensemble/mocks.py` -- scene z threaded into fit-kernel resolution
- tests: test_psf.py rfft2 gate (additive), TestFoldingThresholdTiers, TestRomanKernelConventions

### CODE REVIEW

File: kl_pipe/observation.py [hunk 1]
```diff
@@ -618,6 +618,31 @@ def _fine_pixel_response_fft(fine_image_pars, coarse_pixel_scale):
...
+def _validate_psf_kernel_size_args(
+    psf_kernel_size: Optional[int], psf, rc_was_default: bool
+) -> None:
+    """Guard psf_kernel_size usage in the obs builders.
+
+    A pinned kernel size is meaningless without a PSF, and it is defined in
+    fine-scale pixels (pixel_scale / oversample), so it must be tied to an
+    explicitly chosen render_config: builder-default configs are rebuilt by
+    ``InferenceTask`` at a priors-derived oversample via
+    ``with_render_config``, which does not carry the pinned size and would
```
[the remark this answers is not in the document]
Comment (sweverett) [reply kl_pipe/observation.py]: Confirmed real: InferenceTask.from_obs (task.py:216-218) rebuilds explicit-rc grism obs to fill line_window_halfwidth on the analytic path, dropping the pin exactly on the production fit path. Fixed in 91f104f: psf_kernel_size is now a field on ImageObs/GrismObs (pytree aux), preserved by with_render_config on both classes. Regression test added (test_with_render_config_preserves_pinned_kernel). The builder guard still blocks pinned+default-rc, since a default-rc rebuild changes oversample and the fine-pixel pin would denote a different angular size.

File: kl_pipe/ensemble/spec.py [hunk 2]
```diff
@@ -44,17 +44,118 @@ def _reject_unknown(d: dict, allowed: Tuple[str, ...], context: str) -> None:
...
+            )
+
+
+def _parse_roman_wfi_psf(block: dict, context: str) -> PSFSpec:
+    _reject_unknown(block, ('type', 'sca', 'pupil_bin'), context)
+    return PSFSpec(
+        psf_type='roman_wfi',
+        sca=int(block.get('sca', _ROMAN_WFI_DEFAULT_SCA)),
+        pupil_bin=int(block.get('pupil_bin', _ROMAN_WFI_DEFAULT_PUPIL_BIN)),
+    )
```
[the remark this answers is not in the document]
Comment (sweverett) [reply kl_pipe/ensemble/spec.py:340]: Fixed in 91f104f: _require_yaml_int rejects floats/strings/bools with a clear error instead of coercing. Test added (test_non_integer_sca_pupil_bin_rejected).

File: kl_pipe/ensemble/mocks.py [hunk 3]
```diff
@@ -62,9 +68,137 @@ def _build_gaussian_psf(fwhm_arcsec: float):
...
+        short-wavelength pupil image.
+    """
+    _import_galsim()
+    from galsim import roman
+
+    key = (int(sca), int(pupil_bin), round(float(wavelength_nm), 1), bandpass)
+    if key not in _ROMAN_PSF_CACHE:
+        _ROMAN_PSF_CACHE[key] = roman.getPSF(
+            SCA=key[0], bandpass=bandpass, wavelength=key[2], pupil_bin=key[1]
+        )
```
[the remark this answers is not in the document]
Comment (sweverett) [reply kl_pipe/ensemble/mocks.py]: Fixed in 91f104f: FIFO cap at 16 entries (within-fit reuse needs a handful; cross-fit wavelengths never repeat). Test added (test_roman_psf_cache_bounded).

File: tests/test_roman_psf.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+REPO_ROOT = Path(__file__).resolve().parent.parent
+REGISTRY = REPO_ROOT / 'configs' / 'observing'
+DEV_SPEC = REPO_ROOT / 'configs' / 'ensembles' / 'sigma_eps_cosi_dev.yaml'
+
+# canonical geometry (matches the observing-config registry)
+PIXEL_SCALE = 0.11  # arcsec/pix
+
+# ensemble z range endpoints (specs draw z uniform on [1.0, 1.9])
+Z_LO = 0.55
+Z_HI = 1.9
```
[the remark this answers is not in the document]
Comment (sweverett) [reply tests/test_roman_psf.py]: Comment fixed in 91f104f: Z_HI = current spec draw upper bound (pinning reference); Z_LO = 0.55 is the planned ensemble lower bound (Halpha enters G150 at z~0.52), kept as the low-z endpoint to exercise the full kernel-size spread.

**Comment by sweverett** [comment]
Round 3 (user-directed), 42799fc:

- canonical_P_roman.yaml now sets folding_threshold: 0.01 (fit kernels only). Gate: dual-fidelity MAP-bias experiment (truth rendered with default kernels, fit with loosened; noiseless; delta from a matched-kernel control) -- ft=0.01 passes with all parameter shifts <= 0.07 sigma and shear < 0.01 sigma; ft=0.02 fails (continuum -0.5 sigma, cosi -0.37 sigma: truncated wing flux renormalizes into a stamp-wide pedestal the continuum parameter absorbs).
- Mock-vs-fit kernel fidelity split is first-class: PSFSpec.mock_folding_threshold, default None = galsim 5e-3, so mock data always renders at least as accurately as the fit model (validated, raises otherwise).
- Flagship migrated to Roman WFI PSFs with the same dual fidelity; broadband synthetic data moved to the galsim-native backend (the scipy backend's kernel-on-profile-grid convolution cannot host the Roman far wings). Reruns pass: dev joint Nsigma 0.178 (297 s), production 0.735 (642 s; ~1.26x the gaussian runtime, matching the ft=0.01 projection). All 21 production parameters recover within ~1 sigma; fresh plot set in tests/out/00_flagship/.
- Diagnostic figure test (test_folding_threshold_diagnostic_plot) regenerates the cost-vs-fidelity panels + residual maps under make diagnostics.

**Comment by sweverett** [comment]
Diagnostics for the addendum commits (images hosted on `cc/pr58-assets`, figures-only branch, not for merge; all regenerate under tests/out/).

rfft2 swap, old vs new path. Test kernel is a deliberately asymmetric aberrated OpticalPSF (not the roman kernel) so orientation bugs can't hide. Right panel |new-old|, log scale; max rel delta 4.0e-16 value, 8.0e-16 grad.

rfft2 swap

Kernel conventions with the real roman kernel: k-space fused path vs galsim native (top, 7.0e-4 max|diff|/peak; transposed control 8.0e-3), real-space grism conv vs galsim fine reference (bottom).

kernel conventions

PSF size and ellipticity across the 18 SCAs at H158 eff lambda; dashed = SCA 10 (pinned). ~3% size spread, |e| to ~0.05; single-SCA caveat noted in the addendum.

sca scan

Chromatic (11-pt bandpass-weighted) vs monochromatic kernel at H158: d_sigma/sigma = +0.09%.

chromatic vs mono

File: kl_pipe/observation.py [hunk 5]
```diff
@@ -618,6 +630,34 @@ def _fine_pixel_response_fft(fine_image_pars, coarse_pixel_scale):
...
+        return
+    if psf is None:
+        raise ValueError("psf_kernel_size given without a psf")
+    if rc_was_default:
+        raise ValueError(
+            "psf_kernel_size requires an explicit render_config; the "
+            "builder-default render_config is rebuilt by InferenceTask via "
+            "with_render_config, which would silently drop the pinned "
+            "kernel size"
+        )
```
[the remark this answers is not in the document]
Comment (sweverett) [reply kl_pipe/observation.py:660]: Reworded the guard message to state the real failure (preserved pixel count at a different fine pixel scale, hence a different angular size) in 97e7b57.

File: kl_pipe/observation.py [hunk 6]
```diff
@@ -663,11 +704,19 @@ def build_image_obs(
         priors-sized rc and rebuild the obs internally. For bespoke
         (non-inference) rendering with tight priors, pass an explicit
         ``build_image_render_config(...)`` result.
+    psf_kernel_size : int, optional
+        Explicit PSF kernel stamp size in fine-scale pixels
+        (``pixel_scale / oversample``), passed to ``precompute_psf_fft``.
+        Pins the kernel array shape (odd, >= automatic good size; raises
+        otherwise). Requires an explicit ``render_config``: the
+        builder-default config is rebuilt by ``InferenceTask`` via
+        ``with_render_config``, which would silently drop the pinned size.
```
[the remark this answers is not in the document]
Comment (sweverett) [reply kl_pipe/observation.py]: Aligned the docstring with the oversample/angular-size mismatch in 97e7b57.

File: kl_pipe/psf.py [hunk 7]
```diff
@@ -41,7 +41,7 @@
 class PSFData:
     """Pre-computed PSF arrays for JAX FFT convolution."""
 
-    kernel_fft: jnp.ndarray  # pre-FFT'd kernel (padded)
+    kernel_fft: jnp.ndarray  # pre-rFFT'd kernel (padded); shape (P, P//2+1)
```
[the remark this answers is not in the document]
Comment (sweverett) [reply kl_pipe/psf.py]: Fixed the comment to (Nrow_pad, Ncol_pad//2 + 1) for the rectangular rfft2 case in 97e7b57.

**Comment by sweverett** [comment]
Dual-fidelity kernel bias vs z -- the accuracy case for the folding_threshold tier schedule in one figure. Left: per-parameter MAP bias of ft=0.01 fit kernels against an exact-kernel control (noiseless, canonical_P). The shape/shear/continuum block grows with kernel size and cosi crosses the 0.07 sigma envelope at z=1.9. Right: the ratified schedule (ft=0.01 below z=1.2, ft=5e-3 above) -- above z=1.2 the fit kernel is bit-identical to the mock kernel, so the mismatch bias is exactly zero.

dual fidelity bias vs z
